envision/web/src/components/components.js

```diff
@@ -0,0 +1,343 @@
+  for (const [_, mesh] of Object.entries(vehicleMeshTemplates)) {
+    if (mesh == null) {
+      return false;
+    }
+  }
+  return true;
+};
+
+// Vehicles and agent labels
+export function Vehicles({ scene, worldState, vehicleRootUrl, egoView }) {
+  if (scene == null) {
+    return null;
+  }
+
+  const vehicleMeshTemplatesRef = useRef({});
+  const vehicleMeshesRef = useRef({});
+  const agentLabelGeometryRef = useRef({});
+
+  let vehicleMeshTemplates = vehicleMeshTemplatesRef.current;
+  let vehicleMeshes = vehicleMeshesRef.current;
+  let agentLabelGeometry = agentLabelGeometryRef.current;
+
+  // Load mesh asynchronously
+  useEffect(() => {
+    for (const [vehicleFilename, meshTemplate] of Object.entries(
+      vehicleMeshTemplates
+    )) {
+      if (meshTemplate != null) {
+        continue;
+      }
+
+      SceneLoader.ImportMesh(
+        "",
+        vehicleRootUrl,
+        vehicleFilename,
+        scene,
+        (meshes) => {
+          let rootMesh = meshes[0];
+          rootMesh.isVisible = false;
+          let rootMeshMin = new Vector3();
+          let rootMeshMax = new Vector3();
+          let childMeshes = rootMesh.getChildMeshes();
+          for (let i = 0; i < childMeshes.length; i++) {
+            let child = childMeshes[i];
+            child.isVisible = false;
+
+            let material = new StandardMaterial(
+              `material-${vehicleFilename}`,
+              scene
+            );
+            material.backFaceCulling = false;
+            material.specularColor = new Color3(0, 0, 0);
+            if (child.material) {
+              // Currently only use flat shading, replace imported pbr material with standard material
+              material.id = child.material.id;
+              material.diffuseColor = child.material.albedoColor;
+            }
+            child.material = material;
+
+            let childMin = child.getBoundingInfo().boundingBox.minimumWorld;
+            let childMax = child.getBoundingInfo().boundingBox.maximumWorld;
+
+            if (i == 0) {
+              rootMeshMin = childMin;
+              rootMeshMax = childMax;
+            } else {
+              rootMeshMin = Vector3.Minimize(rootMeshMin, childMin);
+              rootMeshMax = Vector3.Maximize(rootMeshMax, childMax);
+            }
+          }
+
+          rootMesh.setBoundingInfo(new BoundingInfo(rootMeshMin, rootMeshMax));
+
+          vehicleMeshTemplates[vehicleFilename] = rootMesh;
+        }
+      );
+    }
+    // This useEffect is triggered when the vehicleMeshTemplate's keys() change
+  }, [Object.keys(vehicleMeshTemplates).sort().join("-")]);
+
+  useEffect(() => {
+    if (!meshesLoaded(vehicleMeshTemplates)) {
+      return;
+    }
+
+    let nextVehicleMeshIds = new Set(Object.keys(worldState.traffic));
+    let vehicleMeshIds = new Set(Object.keys(vehicleMeshes));
+
+    let vehicleMeshIdsToRemove = difference(vehicleMeshIds, nextVehicleMeshIds);
```

Comment:
Not for this PR: This diffing technique could be abstracted out into a helper or other structure for later reuse.